### PR TITLE
buildRustPackage: build cargoDeps with `pname`

### PR DESCRIPTION
**Pull request by limeytexan**
Using the requesting package's `name` as an input to `fetchCargoTarball()`
has the effect of changing the resulting path of `cargoDeps` when version
numbers change, which in turn leads to a mismatched `cargoSha256` checksum.
Replacing this `name` input with `pname` ensures that differing versions
of the same package will use the same `cargoDeps`. This is useful as it
preserves the ability to change or override the version number for a
package build without requiring a change of checksum.

Addresses #112763

###### Motivation for this change
With this patch in place the reproducer shown in #112763 works as expected:
```
[brantley@clubsoda:~/src/flox-rust-testing]$ cat default.nix 
{ version ? "1.0.0" }:

with import <nixpkgs> { };

rustPlatform.buildRustPackage rec {
  pname = "flox-rust-testing";
  inherit version;
  src = lib.cleanSource ./.;
  cargoSha256 = "109c8q10fib9rihy6lxbcmfjc28ys5yjzkla7z9xa2x1vxylp1r4";
}

[brantley@clubsoda:~/src/flox-rust-testing]$ nix-build -I nixpkgs=/tmp/nixpkgs
/nix/store/qf9d57qgwwqv40sya69063978lhb8akd-flox-rust-testing-1.0.0

[brantley@clubsoda:~/src/flox-rust-testing]$ nix-build -I nixpkgs=/tmp/nixpkgs --argstr version "1.2.3"
/nix/store/vvb6iln3n73l5fy3ykil992r9j1qp3dv-flox-rust-testing-1.2.3

[brantley@clubsoda:~/src/flox-rust-testing]$ 
```

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by danieldk — CHANGES_REQUESTED** [review]
General comments on the up- and downsides of this change:

https://github.com/NixOS/nixpkgs/issues/112763#issuecomment-777417991

More specifically to this PR: this PR should also update the `cargoSha256` of every existing derivation that uses `buildRustPackage`, right?

**Comment by Mic92** [comment]
I am actually in favor of this given that we have https://github.com/NixOS/nixpkgs/blob/65aeed59c52c572eccdf9237ce45d9fb0fd18b91/pkgs/build-support/rust/default.nix#L172

to catch user error. Across different packages makes more sense to have it.

**Comment by Mic92** [comment]
> General comments on the up- and downsides of this change:
> 
> [#112763 (comment)](https://github.com/NixOS/nixpkgs/issues/112763#issuecomment-777417991)
> 
> More specifically to this PR: this PR should also update the `cargoSha256` of every existing derivation that uses `buildRustPackage`, right?

Out-of-tree users might get also unhappy because their packages break and checksums are incompatible between unstable and stable, so we need to do this carefully.

**Comment by danieldk** [comment]
> I am actually in favor of this given that we have
> 
> https://github.com/NixOS/nixpkgs/blob/65aeed59c52c572eccdf9237ce45d9fb0fd18b91/pkgs/build-support/rust/default.nix#L172
> 
> to catch user error. Across different packages makes more sense to have it.

If we make such a breaking change, maybe we should remove dependence on `pname` as well?

File: pkgs/build-support/rust/default.nix
Comment (grahamc) [inline pkgs/build-support/rust/default.nix:64]: Not every package will have a pname, we can fall back to the `name` if it isn't set:

 ```suggestion
        name = args.pname or args.name;
```

Example:

```
nix-repl> args = { name = "foo"; }

nix-repl> args.pname or args.name  
"foo"
```

**Comment by limeytexan** [comment]
> More specifically to this PR: this PR should also update the `cargoSha256` of every existing derivation that uses `buildRustPackage`, right?

Yes, thanks for highlighting - I'll update the PR accordingly.  :+1: 

> Out-of-tree users might get also unhappy because their packages break and checksums are incompatible between unstable and stable, so we need to do this carefully.

I was chatting with @grahamc about this and we weren't exactly sure what you meant - what specific/extra steps would you recommend? Thanks!

**Comment by danieldk** [comment]
> I was chatting with @grahamc about this and we weren't exactly sure what you meant - @Mic92 what specific/extra steps would you recommend? Thanks!

We had this issue twice before, where there was a change invalidated all `cargoSha256` attributes. This is a problem for third-party repositories:

- Their existing hashes are invalidated when packages are built against nixpkgs-unstable/master.
- Repositories that target both nixpkgs 20.09 and nixpkgs-stable, need two `cargoSha256`s, conditional on nixpkgs <= 20.09 and later.

**Comment by limeytexan** [comment]
> We had this issue (at least) twice before, where there was a change invalidated all `cargoSha256` attributes. This is a problem for third-party repositories:
> 
> * Their existing hashes are invalidated when packages are built against nixpkgs-unstable/master.
> * Repositories that target both nixpkgs 20.09 and nixpkgs-stable, need two `cargoSha256`s, conditional on nixpkgs <= 20.09 and later.

Understood - thanks for the background. I see this as a really good reason to avoid patches which trigger a "treewide" recalculation of `cargoSha256` attributes, but is there any thing specifically I need to do in this instance by way of this PR? Sorry if I'm missing something obvious - thanks!

**Comment by danieldk** [comment]
@limeytexan: thanks for the treewide update :+1: , though it currently misses cases where SRI hashes are used. E.g.:

```
❯ nix-build -A broot
[...]
error: --- Error --- nix-daemon
hash mismatch in fixed-output derivation '/nix/store/mv8qlbfw3823rb7x04bv7kjxbpba8p40-broot-vendor.tar.gz.drv':
  specified: sha256-ffFS1myFjoQ6768D4zUytN6F9paWeJJFPFugCrfh4iU=
     got:    sha256-JmBZcDVYJaK1cK05cxx5BrnGWp4t8ca6FLUbvIot67s=
error: --- Error ---------------------------------------------------------------------------------------------------------------- nix-build
path '/nix/store/vdrf94baf9lwydmb7h5a5zs3prpngm1b-broot-vendor.tar.gz' is not valid
```

Seems like the only difference is that it needs `(wanted|specified)` as the pattern in place of `wanted`.

**Comment by limeytexan** [comment]
Thanks @danieldk - I'll update and get back to you!

**Comment by danieldk** [comment]
> I see this as a really good reason to avoid patches which trigger a "treewide" recalculation of cargoSha256 attributes, but is there any thing specifically I need to do in this instance by way of this PR? Sorry if I'm missing something obvious - thanks!

I think it only has a bearing on *whether* we want to make this change and if so, when (perhaps it is better timed just before the next release), rather than any modifications to the PR itself.

Pinging @jonringer as one of the release managers of 21.05.

**Comment by limeytexan** [comment]
> Also, if we are breaking all Cargo vendoring hashes anyway, we should probably decide if we want to use `pname` at all. It's a bit strange that the output is dependent on the name at all, rather than just the content hash, and removing `pname` would also bring it in line with the defaults of some other frequently-used fetchers (`fetchzip, fetchFromGitHub, fetchFromGitLab, etc.).

That's actually where my effort started - by dropping `name` from the `fetchCargoTarball()` invocation attrset entirely - but then the discussion in  #85703 seemed to indicate that having the "name" in the package was in general preferable to having them all called `cargo-deps-vendor.tar.gz`. With that I thought that replacing `name` with `pname` would be the smallest change to achieve what I need, and figured that this was the proposal that would be most likely to succeed. I don't actually mind which way to go, but I do need to know which way it _will_ go because I'd like to get on with rebuilding our copy of nixpkgs with this update.

If you would like me to update the patch once again to drop the `name` entirely just say the word and I'll regenerate those checksums once again while the process is fresh in my mind!  :+1:

**Comment by Infinisil** [comment]
I think we should introduce a solution that allows us to do changes to the hash calculation without always all hashes. A seemingly good way to do this is to introduce a `cargoTarballHashVersion` (defaulting to `0`) argument to `buildRustPackage`. We can then condition any change of hash calculation to only happen when `cargoTarballHashVersion` is above a certain value. In this case we would make the `name` passed to `fetchCargoTarball` be different for when `cargoTarballHashVersion >= 1`. This then doesn't break any existing hashes, doesn't need any treewide updates, and still allows users to specify `cargoTarballHashVersion = 1` when calling `buildRustPackage` for the new behavior.

**Comment by limeytexan** [comment]
I like @Infinisil's suggestion to implement a `cargoTarballHashVersion` feature after or in conjunction with this patch - in other words I don't think we can avoid this one last "treewide" recalculation.

In the meantime I've modified the patch as suggested to drop the `name` altogether (instead of replacing it with `pname`), rebased it against the latest `nixpkgs:master` branch and recalculated the checksums once again. Please let me know when you expect to merge the patch so that I can be ready to rebase and recalculate checksums one last time before the merge. Thanks!

**Comment by danieldk** [comment]
> I like @Infinisil's suggestion to implement a `cargoTarballHashVersion` feature after or in conjunction with this patch - in other words I don't think we can avoid this one last "treewide" checksum recalculation.

Alternatively, we could be to have an optional attribute `cargoDepsName` or `cargoName` in `buildRustPackage`, which still defaults to `${pname}-${version}`. Then no hashes have to be updated, and you could still address #112763 by setting `cargoDepsName = name` in your derivations. Then we could flip the switch tree-wide at our convenience.

I see the benefits of versioning in the future, on the other hand it will also make the interface more complicated. `cargoDepsName` would just use an existing argument attribute of `fetchCaroTarball`.

**Comment by Mic92** [comment]
> `cargoTarballHashVersion`

I second this. The name is also quite long.

**Comment by veprbl** [comment]
> but then the discussion in #85703 seemed to indicate that having the "name" in the package was in general preferable to having them all called `cargo-deps-vendor.tar.gz`

This interpretation is wrong. Consider a following example:

```nix
with import <nixpkgs> {};
  
[
  (fetchurl {
    name = "foo-1.2.3.tar.gz";
    url = "https://example.com";
    sha256 = "sha256-6o+sfGX7WJsNU1YPUlH3T56bJDR43Laz6nm142RJyNk=";
  })
  (fetchurl {
    name = "cargo-deps-vendor.tar.gz";
    url = "https://example.com";
    sha256 = "sha256-6o+sfGX7WJsNU1YPUlH3T56bJDR43Laz6nm142RJyNk=";
  })
]
```

Filename in the store has no effect on the hash, and this is why nobody suggests to wipe the names in the store paths.

**Comment by Infinisil** [comment]
@veprbl No it definitely does, the name influences the store hash. You probably just didn't change the hash between your examples.

**Comment by veprbl** [comment]
> @veprbl No it definitely does, the name influences the store hash.

Not for flat hashes

**Comment by Infinisil** [comment]
Oh I misunderstood. I thought you meant "the filename [of the derivation] has no effect on the hash [in the store path]", but you meant that the filename [the store path] has no effect on the hash [the sha256 hash].

**Comment by danieldk** [comment]
I have made a PR with the alternative proposal with `cargoDepsName` in #113176.

**Comment by limeytexan** [comment]
> I have made a PR with the alternative proposal with `cargoDepsName` in #113176.

Thanks @danieldk - I see that's been merged so we can close this out. Thanks everyone for your help!

**Pull request closed without merging**
